src/main/java/neura/ui/Neura.java

```diff
@@ -0,0 +1,277 @@
+            printTaskStatus("Nice! I've marked this task as done:", taskIndex);
+        } else {
+            throw new NeuraException("Invalid task number.");
+        }
+    }
+
+    private void unmarkTaskAsDone(String userInput) throws NeuraException {
+        int taskIndex = getTaskIndex(userInput);
+        if (isValidTaskIndex(taskIndex)) {
+            tasks.get(taskIndex).markAsNotDone();
+            printTaskStatus("OK, I've marked this task as not done yet:", taskIndex);
+        } else {
+            throw new NeuraException("Invalid task number.");
+        }
+    }
+
+    private int getTaskIndex(String userInput) throws NeuraException {
+        try {
+            return Integer.parseInt(userInput.split(" ")[1]) - 1;
+        } catch (NumberFormatException | ArrayIndexOutOfBoundsException e) {
+            throw new NeuraException("Invalid task number format.");
+        }
+    }
+
+    private boolean isValidTaskIndex(int index) {
+        return index >= 0 && index < tasks.size();
+    }
+
+    private void printTaskStatus(String message, int taskIndex) {
+        System.out.println(SEPARATOR);
+        System.out.println(message);
+        System.out.println(tasks.get(taskIndex));
+        System.out.println(SEPARATOR);
+    }
+
+    private void addTodoTask(String userInput) throws NeuraException {
+        // Ensure that the description after the "todo" command is not empty
+        if (userInput.trim().equals(TODO_COMMAND)) {
+            throw new NeuraException("The description of a todo cannot be empty.");
+        }
+
+        // Extract the description after the "todo" command
+        String description = userInput.substring(TODO_COMMAND.length()).trim();  // Skip "todo" and trim extra spaces
+
+        if (description.isEmpty()) {
+            throw new NeuraException("The description of a todo cannot be empty.");
+        }
+
+        tasks.add(new Todo(description));  // Add a new Todo task to the list
+        printTaskAddedMessage();
+        saveTasks();  // Save tasks to file after adding
+    }
+
+
+    private void addDeadlineTask(String userInput) throws NeuraException {
+        try {
+            String taskDetails = userInput.substring(9).trim();
+            String[] details = taskDetails.split(" /by ", 2);
+            if (details.length < 2) throw new NeuraException("Invalid deadline format. Use: deadline [task] /by [time]");
+            tasks.add(new Deadline(details[0], details[1]));
+            printTaskAddedMessage();
+            saveTasks();  // Save tasks to file after adding
+        } catch (IndexOutOfBoundsException e) {
+            throw new NeuraException("Invalid deadline format. Use: deadline [task] /by [time]");
+        }
+    }
+
+    private void addEventTask(String userInput) throws NeuraException {
+        try {
+            String taskDetails = userInput.substring(6).trim();
+            String[] details = taskDetails.split(" /from ", 2);
```

Comment:
consider extracting low-level string operations to separate methods to improve SLAP